### PR TITLE
Client v2 filter

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: synapse/api/filtering.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            pass  # format is optional
+
+        try:
+            event_select_list = definition["select"]
+            for select_key in event_select_list:
+                if select_key not in ["event_id", "origin_server_ts",
+                                      "thread_id", "content", "content.body"]:
+                    raise SynapseError(400, "Bad select: %s" % (select_key,))
+        except KeyError:
+            pass  # select is optional
```
Comment (erikjohnston) [inline synapse/api/filtering.py]: We've changed from testing if a key exists further up to catching KeyError.

File: synapse/api/filtering.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+        if ("bundle_updates" in definition and
+                type(definition["bundle_updates"]) != bool):
```
Comment (erikjohnston) [inline synapse/api/filtering.py]: Line continuation indent?

File: synapse/rest/client/v2_alpha/filter.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        if target_user != auth_user:
+            raise AuthError(403, "Cannot get filters for other users")
+
+        if not self.hs.is_mine(target_user):
+            raise SynapseError(400, "Can only get filters for local users")
+
+        try:
+            filter_id = int(filter_id)
+        except:
+            raise SynapseError(400, "Invalid filter_id")
```
Comment (erikjohnston) [inline synapse/rest/client/v2_alpha/filter.py]: We're using ints as filter ids?

File: synapse/storage/filtering.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+from twisted.internet import defer
+
+from ._base import SQLBaseStore
+
+import json
+
+
+# TODO(paul)
+_filters_for_user = {}
```
Comment (erikjohnston) [inline synapse/storage/filtering.py]: What is this?

File: synapse/storage/filtering.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            sql = (
+                "SELECT MAX(filter_id) FROM user_filters "
+                "WHERE user_id = ?"
+            )
+            txn.execute(sql, (user_localpart,))
+            max_id = txn.fetchone()[0]
+            if max_id is None:
+                filter_id = 0
+            else:
+                filter_id = max_id + 1
```
Comment (erikjohnston) [inline synapse/storage/filtering.py]: Why are we not using auto incrementing row ids?

**Comment by erikjohnston** [comment]
LGTM

**Comment by NegativeMjark** [comment]
> We're using ints as filter ids?

Apparently so

> Why are we not using auto incrementing row ids?

So that each user gets their own set of auto-incrementing filter IDs I guess.

**Pull request merged**